storage/client_raft_test.go

```diff
@@ -813,3 +815,52 @@ func TestRangeDescriptorSnapshotRace(t *testing.T) {
 		}
 	}
 }
+
+// TestReplicateAddAndRemoveRemove verifies multiraft.state remove the
+// remote node correctly when range was removed from the store.
+func TestReplicateAddAndRemoveRemove(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	mtc := startMultiTestContext(t, 3)
+	defer mtc.Stop()
+
+	// Make the split.
+	store0 := mtc.stores[0]
+	splitArgs := adminSplitArgs(proto.KeyMin, []byte("b"), proto.RaftID(1), store0.StoreID())
+	if _, err := store0.ExecuteCmd(context.Background(), &splitArgs); err != nil {
+		t.Fatal(err)
+	}
+
+	raftID := proto.RaftID(2)
+	mtc.replicateRange(raftID, 0, 1, 2)
+
+	mtc.unreplicateRange(raftID, 0, 2)
+	mtc.unreplicateRange(raftID, 0, 1)
+
+	rng, err := mtc.stores[1].GetRange(raftID)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// If the remove range action happen before range apply replica config change, the group
+	// will be created again when multiraft received a MsgApp.
+	if err := util.IsTrueWithin(func() bool {
+		return len(rng.Desc().Replicas) == 1
+	}, 1*time.Second); err != nil {
+		t.Fatal(err)
+	}
+
+	// Remove the range from store2.
+	if err = mtc.stores[1].RemoveRange(rng); err != nil {
+		t.Fatal(err)
+	}
+
+	if err = mtc.transport.Send(&multiraft.RaftMessageRequest{proto.RaftID(0), raftpb.Message{
+		From: uint64(mtc.stores[2].RaftNodeID()),
+		To:   uint64(mtc.stores[1].RaftNodeID()),
+		Type: raftpb.MsgHeartbeat,
+	}}); err != nil {
+		t.Fatal(err)
+	}
+	// Wait some time to let multiraft process the MsgHeartbeat.
+	<-time.After(1 * time.Millisecond)
```

Comment:
i don't understand this part - are we just sleeping to convince ourselves that nothing panics? Won't `leaktest.AfterTest(t)` do that anyway as it waits for everything to shut down?